2.train-VAE/betatcvae.py

```diff
@@ -0,0 +1,300 @@
+                 alpha: float = 1.0,
+                 gamma: float = 1.0,
+                 anneal_steps: int = 200,
+                 **kwargs) -> None:
+        super(BetaTCVAE, self).__init__()
+
+        # Initialize class attributes
+        self.latent_dim = latent_dim
+        self.anneal_steps = anneal_steps
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+
+        if hidden_dims is None:
+            hidden_dims = [256, 256, 256]  # Default hidden dimensions if none are provided
+
+        # Build Encoder
+        modules = []
+        in_dim = input_dim
+        for h_dim in hidden_dims:
+            modules.append(
+                nn.Sequential(
+                    nn.Linear(in_dim, h_dim),
+                    nn.ReLU())
+            )
+            in_dim = h_dim
+
+        self.encoder = nn.Sequential(*modules)
+        self.fc1 = nn.Linear(input_dim, 256) 
+        self.fc_mu = nn.Linear(256, latent_dim) 
+        self.fc_logvar = nn.Linear(256, latent_dim)  
+        self.fc2 = nn.Linear(latent_dim, 256)  
+        self.fc3 = nn.Linear(256, input_dim) 
+
+        # Build Decoder
+        self.decoder_input = nn.Linear(latent_dim, 256)  
+        modules = []
+        hidden_dims.reverse() 
+
+        for i in range(len(hidden_dims) - 1):
+            modules.append(
+                nn.Sequential(
+                    nn.Linear(hidden_dims[i], hidden_dims[i + 1]),
+                    nn.ReLU())
+            )
+
+        self.decoder = nn.Sequential(*modules)
+        self.final_layer = nn.Linear(hidden_dims[-1], input_dim)  
+
+    def encode(self, input: torch.Tensor) -> List[torch.Tensor]:
+        result = self.encoder(input) 
+        print(f"Encoder output shape: {result.shape}")
+        result = self.fc1(result)  
+        mu = self.fc_mu(result)  
+        log_var = self.fc_logvar(result)  
+        return [mu, log_var]
+
+    def decode(self, z: torch.Tensor) -> torch.Tensor:
+        result = self.decoder_input(z)  
+        result = self.decoder(result)  
+        result = self.final_layer(result)  
+        return result
+
+    def reparameterize(self, mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor:
+        std = torch.exp(0.5 * logvar)  
+        eps = torch.randn_like(std)  
+        return eps * std + mu  
+
+    def forward(self, x):
+        h1 = F.relu(self.fc1(x))  
+        mu = self.fc_mu(h1)  
+        logvar = self.fc_logvar(h1) 
+        z = self.reparameterize(mu, logvar)  
+        h2 = F.relu(self.fc2(z)) 
+        recon_x = self.fc3(h2)  # Get reconstruction
+        return recon_x, mu, logvar, z
+
+    def log_density_gaussian(self, x: torch.Tensor, mu: torch.Tensor, logvar: torch.Tensor):
+        normalization = -0.5 * (math.log(2 * math.pi) + logvar) 
+        inv_var = torch.exp(-logvar)  
+        log_density = normalization - 0.5 * ((x - mu)**2 * inv_var)  
+        return log_density
+
+    def loss_function(self,
```

Comment:
This is a complex loss function. I think some inline documentation might be useful for others to understand better. Consider adding comments explaining what each term is and what it is doing.

Reply:
yep